video4linux Channel Switching Fix

Porkoo opened the pull request:
This branch adds video4linux device (channel) switching functionality. Some of the code might be a bit unintuitive as `capture_function` needs to halt before `release` is called. Everything seems to work when tested with a Datapath VisionRGB-E2S, but code revisions and suggestions would be appreciated.

leikareipa commented:
Thanks! Looks good.

Does `capture_function()` need to stop when changing input channels? What if it instead does a `push_capture_event(capture_event_e::no_signal); continue;` where `ioctl(CAPTURE_HANDLE,...)` fails, i.e. where it currently does `push_capture_event(capture_event_e::unrecoverable_error); break;`? Since opening a new channel modifies CAPTURE_HANDLE, the capture function should pick it up automatically, and in the meantime ioctl() should fail. In theory, anyway - I don't have a multi-channel Linux-supported card to test on.

I'd modify `get_device_maximum_input_count()` so that instead of returning as soon as a non-available video device is found, it counts the number of devices it was able to open. According to the Datapath docs, you can customize which /dev/videoX the card appears in (using rgb133_video_nr), so they might be all over the place. Maybe also some way to identify that a particular openable /dev/video is indeed of this VisionRGB device rather than a webcam or something?

Porkoo commented:
Thanks for the feedback! I'll definitely look into fixing `get_device_maximum_input_count()`. If I recall correctly from the Vision drivers, there is a way to check whether or not a device is a Datapath card.

As for `capture_function()`, I'm not entirely sure. From what I could figure out `capture_function()` would crash the application after the handle was released due to being in the middle of execution.

Porkoo commented:
Great! Looks like replacing `unrecoverable_error` with `invalid_signal` resolved the crashing issue!

leikareipa commented:
My guess is the crash would've been in the part of `capture_function()` that copies the frame's data to local memory. There's a mutex protecting that code, so maybe the function that attempts to change the input channel would also need to be tied to that mutex if no other precautions are taken.

But yeah, since the ioctl() call comes before the frame-copying, handling its failure would prevent the frame copy without involving a mutex. Technically it's not super good coding to 'guard' the thing like this, since at some point somebody could just reorganize the code and now it no longer does the guarding, but the Linux capture code is experimental anyway.

leikareipa commented:
Actually, I wonder whether the fact that you didn't get a crash after the change wasn't just luck? Since passing the ioctl() with a success doesn't mean CAPTURE_HANDLE will still be valid for the rest of the loop's body. Not sure how thread syncing is handled by these file operations.

I'm thinking that to have some guarantee of no crash, we'd need to lock `captureMutex` for the block starting at `// Poll the capture device for a new frame.`, and acquire the mutex when changing the capture channel. Not sure how badly locking the mutex at this point in the capture function will affect performance though.

Porkoo commented:
Okay, I've changed `get_device_maximum_input_count()` to properly count all Vision devices. I've also modified `initialize_hardware()` to support Vision devices that are spread apart or starting at an offset other than 0.

I agree that we can't guarantee stability without some sort of mutex locking, although I'm not entirely sure about how it should be implemented.

leikareipa commented:
Thanks for the changes!

To prevent the `capture_function()` thread syncing from getting hairy, we could simplify it a bit. Let's say that on Linux, since support for it in VCS is experimental, the user can't change the input channel at runtime, they can only set it via the command-line parameter. So now we don't need to worry about the capture function, we just point /dev/video to the user-provided `INPUT_CHANNEL_IDX` in `initialize_hardware()`, and that's the only time the function runs.

Later on, if need be, we can think about how to organize the threading to allow for runtime changes.

Porkoo commented:
Sounds good! I'll revert my changes on this branch and modify `initialize_hardware()` function to use the user-provided `INPUT_CHANNEL_IDX`. In this context, should `INPUT_CHANNEL_IDX` just be the suffix for `/dev/video`?

leikareipa commented:
`capture_function()` and `set_input_channel()` could be reverted to how they were originally, i.e. there's no way for the program to change the input channel after the hardware has been initialized.

The assumption would be that the user knows which input channel is correct, so I suppose `/dev/video` + `INPUT_CHANNEL_IDX` would suffice (the command-line value is in the range [1,n], and gets converted to [0,n-1] for `INPUT_CHANNEL_IDX`). Going by memory I think the code handles a failure to start capturing on that device, so it should be all good.

Porkoo commented:
Okay! The new version I pushed reverts most of the changes and makes `initialize_hardware()` rely on `INPUT_CHANNEL_IDX`. One thing to note is that I used `INPUT_CHANNEL_IDX - 1` for opening the video device, allowing the `INPUT_CHANNEL_IDX` to match the `/dev/video` suffix.

Porkoo commented:
I also removed my changes to `get_device_maximum_input_count()` as they're not currently necessary, but I can add them back in if necessary.

leikareipa commented:
I'd say just use the raw `INPUT_CHANNEL_IDX` value for now. We should assume the user provides a 1-indexed value on the command-line and that `INPUT_CHANNEL_IDX` is this value converted to 0-index. The way the code is right now, it wouldn't be possible to set /dev/video0 as far as I can see, since it gets assigned a minimum of 0+1.

I don't agree with removing `get_device_maximum_input_count()`, unless you know there was something wrong with it. The About dialog in the GUI reports this value for the user's benefit - or can do if the function returns the correct value. So even if the code doesn't directly use it, it can be useful to have.

I can make any adjustments to the code as I see fit, but I can't test the underlying implementation, since I don't have a multi-channel card on Linux. So it's better if you do the base implementation first and test it out.

Porkoo commented:
Alright! Sorry for all the confusion, I'll push the changes shortly.

leikareipa commented:
Thanks, no worries. I'll squash the commits into one before merging, so it doesn't matter if there's a bunch of smaller things before.

leikareipa commented:
Going back theoretically to the threading issue, I wonder if it wouldn't be a good idea to introduce the concept of a capture channel (a class or whatever), with each capture channel representing a hardware input channel (/dev/videoX) and encapsulating things like `capture_function()`.

On program launch, we'd start up a capture channel for each of the capture card's input channels, so that all capture functions are always running in the background, each in their own thread. Each capture channel would be associated with a boolean of whether the channel is active or not - if not active, its `capture_function()` spins in place with a cooldown timer (e.g. 500 ms, shouldn't be too much of a resource hog); otherwise it actively monitors for frame data. In VCS's thread (capture_api_video4linux.cpp), we'd maintain a list of these capture channels, and have a pointer to the currently active one - basically a pointer variant of `CAPTURE_HANDLE` - so that we can direct queries like `get_resolution()` to the correct channel.

This way, to change the input channel, we'd just have to set that channel's active boolean to true and the others to false, and update our local channel pointer - and maybe fire up a capture_event::new_video_mode or whatever. There'd be a small period of one frame or so during which you could end up with two simultaneous memcpys into the main thread's frame buffer or something like that, but I don't immediately see that as being fatal, the code already ignores the first frame or two after a new video mode is set to avoid displaying garbage.

Not sure how much this would e.g. increase hardware load - maybe having each channel open would force the hardware to constantly push out frames on all channels that are receiving a signal even though you're not reading from them.

I might see if I can get something like this going and then call on you to test it, but we'll see.

Porkoo commented:
Yeah, I'll stay in touch if you need me to test anything. With the solution I had earlier [on this commit](https://github.com/Porkoo/vcs/tree/e6de18e599dd177f7a770756db07ff33352e8d88), I wasn't able to trigger a crash by constantly switching channels, although it seems technically possible (although extremely unlikely) that one could occur. 

I'm not entirely sure if it's a good idea to keep all channels open concurrently since it might cause extra hardware overhead (although I haven't confirmed this). Another concern would be an edge case where someone would want one channel open in VCS and another elsewhere. This would fix a lot of our threading issues, so I'd be willing to do some testing if you can come up with something.

leikareipa commented:
Thanks. Yeah, true - probably not a good idea to keep the channels open. The workaround may simply be to have the channel's capture thread close() its /dev/videoX when it goes inactive and open() it when the channel activates, keeping its thread running at all times. This may also give better separation of concerns, since the main thread may not need to care about the V4L details, just starting the channel threads, hosting the frame buffer, and communicating with the rest of VCS. I'm not a huge fan of trusting luck with app crashes though, so it would be a good thing to resolve.

Porkoo commented:
Yeah, we should definitely come up with a stable and reliable solution over relying on luck. I think running each device in a thread that opens and closes on demand might be our best option right now.

leikareipa commented:
I implemented an input channel object that encapsulates the capture function and some other data, and so changing the current input channel is just "delete this->inputChannel; this->inputChannel = new input_channel(...);" without worrying about thread sync.

If you like you could test it out, just build the current repo and in the program hit Shift + 1 and Shift + 2 to toggle between /dev/video0 and /dev/video1.

Porkoo commented:
I just tested and it seems to be working great! Thanks for resolving that.

leikareipa commented:
No problem, thanks for contributing and testing!

leikareipa commented:
Seen any crashes or freezes when switching between channels yet?

I can produce a freeze by hitting Shift+1 about 2-20 times, i.e. closing the current input channel and opening it again. Also happens when alternating between Shift+1 and Shift+2. Seems it hangs on ioctl() with VIDIOC_QBUF, so could be a driver issue or VCS not initializing/releasing the buffers in the correct way.

Porkoo commented:
I haven't run into any crashed in normal use yet, however I am able to replicate the freeze after pressing Shift+F1 repeatedly for a while.

I am having this error when VCS freezes, which I assume might be related:
`kernel:usercopy: Kernel memory exposure attempt detected from process stack (offset 0, size 12582912)!`

leikareipa commented:
Interesting, thanks. Will just wait for the new driver then and see what happens.

leikareipa commented:
Datapath came out with the new Linux drivers for kernel 5. I wasn't immediately able to get the freeze to happen anymore, so that may be sorted now.

But it seems there's still a memory leak of some kind, as there was with the previous drivers. You can eat up gigs of RAM just by switching between the input channels for a minute or two, or by running/exiting VCS very many times. memcheck doesn't trace it to VCS, so I'm guessing it may be the driver still. Maybe initializing capture in a different way would get around it.

Porkoo commented:
I tested my VisionRGB-E2S with the new drivers this morning and the freeze issue seems to be resolved on my end too. 

Since I've already been keeping the Vision drivers updated to be compatible with Kernel 5.8 and 5.9, I'll look to see if there are any glaring memory issues/workarounds in the Kernel module. It may be something we can't resolve easily without an official fix considering half of the driver is pre-compiled.

leikareipa commented:
Good to hear. I guess I could ask Datapath about the memory issue, just need to reduce it to a minimum reproducible sample. It may also be that using a different kind of V4L capture buffer gets around it.

The leak isn't a huge deal if you regularly reboot the computer and don't constantly switch between channels, but I've also had situations where it eventually fully exhausted the 16 GB I have with nothing else running.

Porkoo commented:
Yeah, I could see the memory leak causing problems with certain applications. Contacting Datapath might be our best bet right now.

leikareipa commented:
I didn't end up contacting Datapath, just used a different kind of V4L capture buffer (V4L2_MEMORY_MMAP instead of V4L2_MEMORY_USERPTR), and I don't see those leaks anymore. The performance of these buffer types seems to be more or less the same, or at least not wildly different. (Also managed to implement dynamic allocation of buffers to the current capture size instead of the maximum possible, which boosted performance by multiples.)
